// react-native-flex-layout/lib/commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Box", {
  enumerable: true,
  get: function () {
    return _Box.default;
  }
});
Object.defineProperty(exports, "Divider", {
  enumerable: true,
  get: function () {
    return _Divider.default;
  }
});
Object.defineProperty(exports, "Flex", {
  enumerable: true,
  get: function () {
    return _Flex.default;
  }
});
Object.defineProperty(exports, "HStack", {
  enumerable: true,
  get: function () {
    return _Stack.HStack;
  }
});
Object.defineProperty(exports, "Selector", {
  enumerable: true,
  get: function () {
    return _Selector.default;
  }
});
Object.defineProperty(exports, "Spacer", {
  enumerable: true,
  get: function () {
    return _Flex.Spacer;
  }
});
Object.defineProperty(exports, "SpacingFuncContext", {
  enumerable: true,
  get: function () {
    return _SpacingFuncContext.SpacingFuncContext;
  }
});
Object.defineProperty(exports, "SpacingFuncProvider", {
  enumerable: true,
  get: function () {
    return _SpacingFuncContext.SpacingFuncProvider;
  }
});
Object.defineProperty(exports, "Stack", {
  enumerable: true,
  get: function () {
    return _Stack.default;
  }
});
Object.defineProperty(exports, "VStack", {
  enumerable: true,
  get: function () {
    return _Stack.VStack;
  }
});
Object.defineProperty(exports, "Wrap", {
  enumerable: true,
  get: function () {
    return _Wrap.default;
  }
});
Object.defineProperty(exports, "defaultSpacingFunc", {
  enumerable: true,
  get: function () {
    return _SpacingFuncContext.defaultSpacingFunc;
  }
});
Object.defineProperty(exports, "even", {
  enumerable: true,
  get: function () {
    return _Selector.even;
  }
});
Object.defineProperty(exports, "firstChild", {
  enumerable: true,
  get: function () {
    return _Selector.firstChild;
  }
});
Object.defineProperty(exports, "getValidChildren", {
  enumerable: true,
  get: function () {
    return _utilities.getValidChildren;
  }
});
Object.defineProperty(exports, "lastChild", {
  enumerable: true,
  get: function () {
    return _Selector.lastChild;
  }
});
Object.defineProperty(exports, "notFirstChild", {
  enumerable: true,
  get: function () {
    return _Selector.notFirstChild;
  }
});
Object.defineProperty(exports, "notLastChild", {
  enumerable: true,
  get: function () {
    return _Selector.notLastChild;
  }
});
Object.defineProperty(exports, "odd", {
  enumerable: true,
  get: function () {
    return _Selector.odd;
  }
});
Object.defineProperty(exports, "select", {
  enumerable: true,
  get: function () {
    return _Selector.select;
  }
});
Object.defineProperty(exports, "useSpacing", {
  enumerable: true,
  get: function () {
    return _SpacingFuncContext.useSpacing;
  }
});
Object.defineProperty(exports, "useSpacingFunc", {
  enumerable: true,
  get: function () {
    return _SpacingFuncContext.useSpacingFunc;
  }
});

var _Box = _interopRequireDefault(require("./Box"));

var _Divider = _interopRequireDefault(require("./Divider"));

var _Flex = _interopRequireWildcard(require("./Flex"));

var _Selector = _interopRequireWildcard(require("./Selector"));

var _SpacingFuncContext = require("./SpacingFuncContext");

var _Stack = _interopRequireWildcard(require("./Stack"));

var _Wrap = _interopRequireDefault(require("./Wrap"));

var _utilities = require("./utilities");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
