// react-native-flex-layout/lib/commonjs/Wrap.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactNative = require("react-native");

var _Flex = _interopRequireDefault(require("./Flex"));

var _Selector = _interopRequireDefault(require("./Selector"));

var _SpacingFuncContext = require("./SpacingFuncContext");

var _utilities = require("./utilities");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

const Wrap = _ref => {
  let {
    spacing = 0,
    shouldWrapChildren = false,
    childrenStyle,
    children,
    ...rest
  } = _ref;
  const spacingValue = (0, _SpacingFuncContext.useSpacing)(spacing);
  const validChildren = (0, _utilities.getValidChildren)(children);
  const clones = !shouldWrapChildren ? validChildren : validChildren.map((child, index) => {
    const key = typeof child.key !== 'undefined' ? child.key : index;
    return /*#__PURE__*/_react.default.createElement(WrapItem, {
      key: key
    }, child);
  });
  return /*#__PURE__*/_react.default.createElement(_Flex.default, _extends({
    mt: -spacing,
    ms: -spacing,
    inline: true,
    wrap: true
  }, rest), /*#__PURE__*/_react.default.createElement(_Selector.default, {
    childrenStyle: [{
      marginTop: spacingValue,
      marginStart: spacingValue
    }, childrenStyle]
  }, clones));
};

var _default = Wrap;
exports.default = _default;

const WrapItem = props => /*#__PURE__*/_react.default.createElement(_reactNative.View, props);
